draft-ietf-ppm-dap.md

```diff
@@ -1391,124 +1320,227 @@ Otherwise, `out` is the pair `(prep_state, prep_msg)`.
+  start(0),
+  continued(1),
+  finished(2),
+  failed(3),
+  (255),
+} PrepareStepState;
 
-This output message for each report in AggregateContinueReq.prepare_steps is
-then sent to the leader in an AggregateContinueResp message, structured as
-follows:
+enum {
+  batch_collected(0),
+  report_replayed(1),
+  report_dropped(2),
+  hpke_unknown_config_id(3),
+  hpke_decrypt_error(4),
+  vdaf_prep_error(5),
+  batch_saturated(6),
+  task_expired(7),
+  unrecognized_message(8),
+  report_too_early(9),
+  (255)
+} ReportShareError;
 
-~~~
 struct {
+  ReportID report_id;
+  PrepareStepState prepare_step_state;
+  select (PrepareStep.prepare_step_state) {
+    case start: ReportShare;
+    case continued: opaque prep_msg<0..2^32-1>;
+    case finished: Empty;
+    case failed: ReportShareError;
+  }
+} PrepareStep;
+
+struct {
+  u16 round;
+  opaque agg_param<0..2^32-1>;
+  PartialBatchSelector part_batch_selector;
   PrepareStep prepare_steps<1..2^32-1>;
-} AggregateContinueResp;
+} AggregationJob;
 ~~~
 
-The order of AggregateContinueResp.prepare_steps MUST match that of the
-PrepareStep values in `AggregateContinueReq.prepare_steps`. The helper's
-response to the leader is an HTTP status code 200 OK whose body is the
-AggregateContinueResp and media type is
-"application/dap-aggregate-continue-resp". The helper then awaits the next
-message from the leader.
+An `AggregationJob` consists of:
 
-[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
+* The current round of the VDAF preparation protocol.
+
+* The opaque, VDAF-specific aggregation parameter provided during the collection
+  flow (`agg_param`),
+
+  [[OPEN ISSUE: Check that this handling of `agg_param` is appropriate when the
+  definition of Poplar is done.]]
+
+* Information used by the Aggregators to determine how to aggregate each report:
+
+    * For fixed_size tasks, the Leader specifies a "batch ID" that determines
+      the batch to which each report for this aggregation job belongs.
+
+      [OPEN ISSUE: For fixed_size tasks, the Leader is in complete control over
+      which batch a report is included in. For time_interval tasks, the Client
+      has some control, since the timestamp determines which batch window it
+      falls in. Is this desirable from a privacy perspective? If not, it might
+      be simpler to drop the timestamp altogether and have the agg init request
+      specify the batch window instead.]
+
+  The indicated query type MUST match the task's query type. Otherwise, the
+  Helper MUST abort with error "queryMismatch".
+
+* The sequence of report shares to aggregate. The `encrypted_input_share` field
+  of the report share is the `HpkeCiphertext` whose index in
+  `Report.encrypted_input_shares` is equal to the index of the aggregator in the
+  task's `aggregator_endpoints` to which the AggregateInitializeReq is being
+  sent.
+
+#### Required methods
+
+##### PUT `/tasks/{task-id}/aggregation_jobs` {#aggregation-job-put}
```

Comment:
* Editorial: It might be helpful to define the resource and message format before getting into the protocol flow.
* Where is the aggregation job ID? We still need a way to link requests for the same job across rounds.

Reply:
Summary of offline discussion: The aggregation job ID is missing here because it has been folded into the redirect URI that the Leader is supposed to post to next to move aggregation forward. (In other words, the aggreagtion job ID is now an implementation detail.) The ideas is to make aggreagtion jobs "asynchronous" the way collection jobs are. However there is an important difference between the two: Here the Helper computes its response immediately, i.e., aggregation is inherently synchronous. Collection is inherently asynchronous because the Leader usually won't be finished computing the aggregate at the time the request is made.

We decided that we should re-align this part with DAP-03, i.e., move the aggregation job ID into the API here.

Reply:
We reached a similar conclusion about the helper's aggregate share resource: the leader should always be able to wait until all relevant aggregation jobs are driven to completion before requesting an aggregate share from the helper. So servicing such a request should just be a matter of collapsing a bunch of output shares into an aggregate share (or, depending on the implementation, collapsing a bunch of partial aggregates over, say, an hour's worth of output shares into the requested aggregate share), which should be doable synchronously for the helper. So I'm also going to move the aggregate share resource back to how it worked in DAP-03 (though we'll still change the request path).

Reply:
> Editorial: It might be helpful to define the resource and message format before getting into the protocol flow.

I'll give that a try after the dust has settled on other changes.

- [ ] put the description of aggregate job endpoints and messages before the prose discussion of the protocol